### ISSUE TITLE
Support armhf

### ISSUE THREAD

**Issue by anthony-cros** [doublethinkco/webthree-umbrella-cross#10]

**Comment by bobsummerwill** [comment]
Do you mean more platforms (ie. build for Sailfish, Ubuntu Phone)?    Or are you specifically thinking ARM vs i386 here?    So the scripts can be used for SBCs.

**Comment by bobsummerwill** [comment]
See https://doublethinkco.wordpress.com/2015/09/22/porting-ethereum-to-mobile-linux/ for target platforms (all ARM).

The list at https://doublethinkco.wordpress.com/ethereum-light-client/ includes the SBCs.    The Intel Edison is the only i386 one for now, though I have an Intel NUC as well, which I would like it to work for.

So in general, I think i386 is very low priority for now.

**Comment by bobsummerwill** [comment]
Hey @phonikg!    Actually, maybe this is a task which would make sense for you to look at?    Or are all your DEVgrant target platforms ARM too?     Have you tried using webthree-umbrella-cross yet?

**Comment by anthony-cros** [comment]
Yes the idea was to explore other architectures like i386 or whichever ones would make sense, i think these scripts can be leveraged easily to do that as they're quite generic already except in a few places like [here](https://github.com/doublethinkco/webthree-umbrella/blob/develop/cross-build/ethereum/boost.sh#L26)
or [here](https://github.com/doublethinkco/webthree-umbrella/blob/develop/cross-build/ethereum/gmp.sh#L20)

**Comment by bobsummerwill** [comment]
So yeah, i386 and amd64 are probably the only ones really worth considering.

Also maybe armhl as well as armel might be needed.  I get a suspicion that either Tizen or Sailfish is actually a different ARM variant than Android.   Will have to dig into that.

Beyond x86 and ARM though?   Balls to them.

**Comment by bobsummerwill** [comment]
armhf, I mean.

**Comment by phonikg** [comment]
Yeah @Bob all EthEmbedded boards are armhf.  I was always thinking on Intel Edison x86 as a bonus but Christoph from Slock.it has that running already no no sense in duplicating that work if they will share it.... my understanding is that they are compiling on device though, so an x86 cross compilation would still be extremely helpful.

**Comment by bobsummerwill** [comment]
Thanks @phonikg.   And you are specifically building armhf for your existing binaries?    Or perhaps that is just happening by default?

**Comment by bobsummerwill** [comment]
Also, @phonikg.

With this existing setup, I've just been able to build myself an ARM binary on a VM in Azure.   So I think we are very, very close to a "done" point, where we are iterating and bug-fixing on something which is functional.

Have you tried the Dockerfile yet yourself?   If not, please do give it a whirl.

just git clone, docker build and then docker run -i -t <hex code> /bin/bash to then do the actual build.

**Comment by anthony-cros** [comment]
@phonikg , you can even just do `docker build .` if you're in the folder containing the Dockerfile :)

**Comment by phonikg** [comment]
While we are all working away at testing, and due to the fact that I have not been able to follow through and get a binary yet at this point, would you be able to provide me with any binaries you guys have been able to create?

Email me a link even, so I could download and try running on a few arm devices?

**Comment by phonikg** [comment]
NVM!! @bobsummerwill posted this an hour ago:(

https://github.com/doublethinkco/webthree-umbrella-cross/releases

**Comment by bobsummerwill** [comment]
@phonikg and @martinbrook:  So it might be a simple as building the xcompiler with arm-unknown-linux-gnueabihf.

This is the other line I was thinking of which @anthony-cros had mentioned as a hack, but might be OK in this instance for armhf ...

    https://github.com/doublethinkco/webthree-umbrella/blob/develop/cross-build/ethereum/boost.sh#L26

Maybe there are more little warts like this?   Or maybe not!

**Comment by anthony-cros** [comment]
> So it might be a simple as building the xcompiler with arm-unknown-linux-gnueabihf

yes I suspect it will be necessary now, in light of all the recent findings

**Comment by bobsummerwill** [comment]
Lots of related ARM ABI discussion happening in https://github.com/doublethinkco/webthree-umbrella-cross/issues/21.

**Comment by anthony-cros** [comment]
Sadly, there doesn't seem to be an explicit option for _armel_ in crosstool-ng:

```
$ ../ct-ng list-samples | grep arm-unknown-linux-gnueabi
	[G..]   arm-unknown-linux-gnueabi
$ ../ct-ng list-samples | grep -i el
	[G..]   m68k-unknown-elf
	[G..]   mips-unknown-elf
	[G.X]   mips64el-n32-linux-uclibc
	[G.X]   mips64el-n64-linux-uclibc
	[G..]   mipsel-sde-elf
	[G..]   mipsel-unknown-linux-gnu
```

**Comment by anthony-cros** [comment]
now i understand we can create such "sample" ourself (some kind of a profile I believe)

**Comment by anthony-cros** [comment]
samples present themselves that way:

```
~/crosstool-ng-1.20.0/samples/arm-unknown-linux-gnueabi$ cat crosstool.config 
CT_LOCAL_TARBALLS_DIR="${HOME}/src"
CT_SAVE_TARBALLS=y
# CT_REMOVE_DOCS is not set
CT_LOG_EXTRA=y
CT_ARCH_FLOAT_SW=y
CT_ARCH_arm=y
CT_KERNEL_linux=y
CT_BINUTILS_LINKER_LD_GOLD=y
CT_BINUTILS_GOLD_THREADS=y
CT_BINUTILS_LD_WRAPPER=y
CT_BINUTILS_PLUGINS=y
CT_LIBC_glibc=y
CT_CC_LANG_CXX=y
# CT_CC_GCC_SJLJ_EXCEPTIONS is not set
CT_DEBUG_dmalloc=y
CT_DEBUG_duma=y
CT_DEBUG_gdb=y
# CT_GDB_CROSS_PYTHON is not set
CT_GDB_NATIVE=y
CT_DEBUG_ltrace=y
CT_DEBUG_strace=y
```
and
```
~/crosstool-ng-1.20.0/samples/arm-unknown-linux-gnueabi$ cat reported.by 
reporter_name="Alexander BIGGA"
reporter_url="http://sourceware.org/ml/crossgcc/2008-06/msg00031.html"
reporter_comment=""
```

**Comment by phonikg** [comment]
Hopefully its as simple as `CT_ARCH_FLOAT_SW=n`

**Comment by anthony-cros** [comment]
I was wondering too, but i doubt it :)

what other differences do you know of between hf/el? can float really be the only one?

**Comment by phonikg** [comment]
I believe its simply software vs hardware based floating point.  This opinion is based mainly on this article:

https://blogs.oracle.com/jtc/entry/is_it_armhf_or_armel

**Comment by anthony-cros** [comment]
worth giving it a shot :)

**Comment by phonikg** [comment]
This is the ct-ng cli gui(oxymoron) pic that shows it as well:

http://www.crosstool-ng.org/_detail/snapshots/002-ctng-target-menu.png?id=start

**Comment by anthony-cros** [comment]
thanks, i played with it a bit but then realized I could simply clone an existing sample and edit the config file directly, it then gets picked up (i'll detail all the steps in a script soon)

**Comment by bobsummerwill** [comment]
If you guys look in #21 you will see ELF header snapshots for multiple scenarios.   Looks like we should fix the NAME tag too, if we can.

**Comment by bobsummerwill** [comment]
Let's bump up to armv7 at the same time.   That screws Rpi, but we can come back to that later.

**Comment by anthony-cros** [comment]
crosstool-ng's _CT_ARCH_ARCH_ seems to hold values like "armv7-a", "armv5te", so I suspect it might address the readelf  _Tag_CPU_arch_ value, but for _Tag_CPU_name_ I'm uncertain (there is a _CT_ARCH_CPU_, but it has values like "arm926ej-s" and "cortex-m3")

**Comment by anthony-cros** [comment]
maybe _CT_ARCH_ARCH_ takes care of both _Tag_CPU_arch_ and _Tag_CPU_name_

**Comment by anthony-cros** [comment]
starting ct-ng build with the following changes:

```
$ diff ~/crosstool-ng-1.20.0/install/lib/ct-ng.1.20.0/samples/arm-unknown-linux-gnueabi/crosstool.config  ~/crosstool-ng-1.20.0/install/lib/ct-ng.1.20.0/samples/arm-unknown-linux-gnueabiel/crosstool.config
5c5
< CT_ARCH_FLOAT_SW=y
---
> CT_ARCH_FLOAT_SW=n
6a7
> CT_ARCH_ARCH=armv7-a
```

**Comment by martinbrook** [comment]
http://randomsplat.com/id192-building-a-hard-float-arm-toolchain.html mentions

CT_ARCH_FPU=”vfpv3″
CT_TARGET_CFLAGS=”-mfloat-abi=hard -mfpu=vfpv3″

**Comment by phonikg** [comment]
Make sure we note that this link references version 1.13 and we are using 1.20

**Comment by phonikg** [comment]
We could always download and run the gui,  then reference the configuration file it produces.

**Comment by anthony-cros** [comment]
@phonikg , are you talking about the ncurse GUI?

**Comment by phonikg** [comment]
Hmm no the crosstool-ng gui ...or are we beyond that already?

**Comment by bobsummerwill** [comment]
Again, back in #21 I added links to the other hacky looking bits in the build process which might need touching.

**Comment by anthony-cros** [comment]
i can confirm that the menu has changed since [this page](http://randomsplat.com/id192-building-a-hard-float-arm-toolchain.html) was written, there is now an explicit option to set hard versus float

my concern is that _CT_ARCH_32=y_ gets removed upon switching from soft to hard, I don't know if it'll have an impact. warning says:

```
.config:152:warning: override: reassigning to symbol ARCH_32
.config:152:warning: override: ARCH_32 changes choice state
```

and I can see it being gone in _crosstool-ng-1.20.0/build/.config_

**Comment by bobsummerwill** [comment]
Right - well I guess we just need to turn the handle and see what comes out, eh!

We can compare with the metadata which @martinbrook dug out in #21 when we have new binaries.

**Comment by phonikg** [comment]
I can confirm that all my SBC's are 32 bit

**Comment by anthony-cros** [comment]
this is what I'm getting with the configuration I mentioned above:

```
xcompiler@d6bab4779159:~/webthree-umbrella/cross-build/ethereum$ readelf -A /home/xcompiler/eth/151116201928/installs/curl/bin/curl
Attribute Section: aeabi
File Attributes
  Tag_CPU_name: "ARM10TDMI"
  Tag_CPU_arch: v5T
  Tag_ARM_ISA_use: Yes
  Tag_THUMB_ISA_use: Thumb-1
  Tag_FP_arch: VFPv2
  Tag_ABI_PCS_wchar_t: 4
  Tag_ABI_FP_rounding: Needed
  Tag_ABI_FP_denormal: Needed
  Tag_ABI_FP_exceptions: Needed
  Tag_ABI_FP_number_model: IEEE 754
  Tag_ABI_align_needed: 8-byte
  Tag_ABI_align_preserved: 8-byte, except leaf SP
  Tag_ABI_enum_size: int
  Tag_ABI_HardFP_use: SP and DP
  Tag_ABI_VFP_args: VFP registers
  Tag_DIV_use: Not allowed
```

a little bit closer it seems... I will try to hardcode the v7 bit in the config file directly

**Comment by bobsummerwill** [comment]
Please could you upload the tgz for that run?  Maybe I can get them running on something?   Maybe @phonikg could!   Thanks!

**Comment by phonikg** [comment]
Yes I could try it right away if you have something.

**Comment by bobsummerwill** [comment]
Ld.so might behave differently on different platforms.   We still don't know how 'compatible binaries' for SO files are determined.   Seems some bad combinations just get 'let through' and crash at runtime!

**Comment by anthony-cros** [comment]
the result with the hardcoding bit:

```
xcompiler@d6bab4779159:~/webthree-umbrella/cross-build/ethereum$ readelf -A /home/xcompiler/eth/151116201928/installs/curl/bin/curl
Attribute Section: aeabi
File Attributes
  Tag_CPU_name: "7-A"
  Tag_CPU_arch: v7
  Tag_CPU_arch_profile: Application
  Tag_ARM_ISA_use: Yes
  Tag_THUMB_ISA_use: Thumb-2
  Tag_FP_arch: VFPv2
  Tag_ABI_PCS_wchar_t: 4
  Tag_ABI_FP_rounding: Needed
  Tag_ABI_FP_denormal: Needed
  Tag_ABI_FP_exceptions: Needed
  Tag_ABI_FP_number_model: IEEE 754
  Tag_ABI_align_needed: 8-byte
  Tag_ABI_align_preserved: 8-byte, except leaf SP
  Tag_ABI_enum_size: int
  Tag_ABI_HardFP_use: SP and DP
  Tag_ABI_VFP_args: VFP registers
  Tag_CPU_unaligned_access: v6
  Tag_DIV_use: Not allowed
```

name and arch now match what @martinbrook reported in https://github.com/doublethinkco/webthree-umbrella-cross/issues/21#issuecomment-159433968

**Comment by anthony-cros** [comment]
i only have a curl binary, I just went for the smallest bit among our components, but I'm now kicking off a full run and will send the result to you guys when it's done (tomorrow morning EST)!

**Comment by bobsummerwill** [comment]
Great!   Thanks, @anthony-cros.   Where will your commits land?  Split across umbrella and umbrella-cross?

**Comment by anthony-cros** [comment]
:(

```
[ 52%] Building CXX object src/jsonrpccpp/CMakeFiles/jsonrpccommon.dir/common/specificationparser.cpp.o
cd /home/xcompiler/eth/151127041039/wd/libjson-rpc-cpp/src/jsonrpccpp && /home/xcompiler/x-tools/arm-unknown-linux-gnueabi/bin/arm-unknown-linux-gnueabi-g++   -Djsonrpccommon_EXPORTS -std=c++11 -isystem/home/xcompiler/eth/151127041039/installs/curl/include -fPIC -I/home/xcompiler/eth/151127041039/wd/libjson-rpc-cpp/catch/src/catch/include -I/home/xcompiler/eth/151127041039/wd/libjson-rpc-cpp/gen -I/home/xcompiler/eth/151127041039/wd/libjson-rpc-cpp/gen/jsonrpccpp/common -I/home/xcompiler/eth/151127041039/src/libjson-rpc-cpp/src/jsonrpccpp/.. -I/home/xcompiler/eth/151127041039/installs/jsoncpp/include -I/home/xcompiler/eth/151127041039/installs/libmicrohttpd/include    -o CMakeFiles/jsonrpccommon.dir/common/specificationparser.cpp.o -c /home/xcompiler/eth/151127041039/src/libjson-rpc-cpp/src/jsonrpccpp/common/specificationparser.cpp
Linking CXX shared library ../../lib/libjsonrpccpp-common.so
cd /home/xcompiler/eth/151127041039/wd/libjson-rpc-cpp/src/jsonrpccpp && /usr/bin/cmake -E cmake_link_script CMakeFiles/jsonrpccommon.dir/link.txt --verbose=1
/home/xcompiler/x-tools/arm-unknown-linux-gnueabi/bin/arm-unknown-linux-gnueabi-g++  -fPIC    -shared -Wl,-soname,libjsonrpccpp-common.so.0 -o ../../lib/libjsonrpccpp-common.so.0.6.0 CMakeFiles/jsonrpccommon.dir/common/errors.cpp.o CMakeFiles/jsonrpccommon.dir/common/specificationwriter.cpp.o CMakeFiles/jsonrpccommon.dir/common/procedure.cpp.o CMakeFiles/jsonrpccommon.dir/common/exception.cpp.o CMakeFiles/jsonrpccommon.dir/common/specificationparser.cpp.o /home/xcompiler/eth/151127041039/installs/jsoncpp/lib/libjsoncpp.a 
/home/xcompiler/x-tools/arm-unknown-linux-gnueabi/lib/gcc/arm-unknown-linux-gnueabi/4.9.1/../../../../arm-unknown-linux-gnueabi/bin/ld.bfd: /home/xcompiler/eth/151127041039/installs/jsoncpp/lib/libjsoncpp.a(json_reader.cpp.o): relocation R_ARM_MOVW_ABS_NC against `__pthread_key_create' can not be used when making a shared object; recompile with -fPIC
/home/xcompiler/eth/151127041039/installs/jsoncpp/lib/libjsoncpp.a: could not read symbols: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [lib/libjsonrpccpp-common.so.0.6.0] Error 1
make[2]: Leaving directory `/home/xcompiler/eth/151127041039/wd/libjson-rpc-cpp'
make[1]: *** [src/jsonrpccpp/CMakeFiles/jsonrpccommon.dir/all] Error 2
make[1]: Leaving directory `/home/xcompiler/eth/151127041039/wd/libjson-rpc-cpp'
make: *** [all] Error 2
```

**Comment by anthony-cros** [comment]
@bobsummerwill , well for armhf, I was hoping only in the ct-ng.sh script, but the above error tells me I'll need to tweak a few more things in the cmake-based scripts :-/

**Comment by anthony-cros** [comment]
https://drive.google.com/file/d/0B4ByIjYZ3FeJMjhCVHlwMndubWc/view?usp=sharing

if you guys want to try the result _curl_ binary at least, and see if it runs properly on an armhf device

**Comment by bobsummerwill** [comment]
The TM1 you have is armhf.   Does it work on that?

**Comment by bobsummerwill** [comment]
I'm still trying to get the RD-PQ flashed and recognized by SDB.   No luck yet.

**Comment by martinbrook** [comment]
[nemo@Jolla ~]$ ./curl
./curl: /lib/libc.so.6: version `GLIBC_2.17' not found (required by ./curl)
[nemo@Jolla ~]$ ls /lib/libc.so.6  -l
lrwxrwxrwx 1 root root 12 2015-04-07 00:09 /lib/libc.so.6 -> libc-2.15.so

Jolla on libc-2.15 looking for 2.17

**Comment by martinbrook** [comment]
Works fine on SailfishOS 2.0 which has 2.19 libc

[nemo@Jolla ~]$ ./curl www.google.com
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>302 Moved</TITLE></HEAD><BODY>
<H1>302 Moved</H1>
The document has moved
<A HREF="http://www.google.co.uk/?gfe_rd=cr&amp;ei=R85YVrT-G86JOvD6scgJ">here</A>.
</BODY></HTML>
[nemo@Jolla ~]$ ls -l /lib/libc-2.19.so 
-rwxr-xr-x 1 root root 992972 2015-05-16 13:38 /lib/libc-2.19.so

**Comment by bobsummerwill** [comment]
Thanks for the update, @martinbrook!

**Comment by bobsummerwill** [comment]
@martinbrook So for Sailfish OS 2.0, /lib/libc.so.6 is a soft-link to /lib/libc-2.19.so, and presumably the loader is doing a >= version check?   Newer is always assumed to be OK.

And also here on Linux, the symbolic references like /lib/libc.so.6 have hard-coded paths to the LIBC6 locations within the executable, right?    Unlike to Windows where it follows a search-path approach to locating DLLs.    So you can always just ship your own DLLs side-by-side with an application and it will prefer those to globally installed ones.

**Comment by phonikg** [comment]
Seems to be working beautifully on Wandboard Quad:

curl --version
curl 7.45.0 (arm-unknown-linux-gnueabihf) libcurl/7.45.0 GnuTLS/3.3.18 zlib/1.2.8 libidn/1.32 libssh2/1.5.0 nghttp2/1.4.0 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets 
debian@debian-armhf:~$ ./curl
curl: try 'curl --help' for more information
debian@debian-armhf:~$ ./curl www.google.com
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>302 Moved</TITLE></HEAD><BODY>
<H1>302 Moved</H1>
The document has moved
<A HREF="http://www.google.ca/?gfe_rd=cr&amp;ei=CO5YVo6gMcyC8Qf355uACQ">here</A>.
</BODY></HTML>
debian@debian-armhf:~$

**Comment by bobsummerwill** [comment]
Thanks, @phonikg!
Maybe you could make a work-branch or something, @anthony-cros?   Where you could commit your WIP on this.    I should get a little time over the weekend and can try to help progress it too.

**Comment by anthony-cros** [comment]
no luck on the TM1:
```
sh-4.1$ ./curl 
sh: ./curl: No such file or directory
```
a message I typically see when I use a "wrong architecture" (like using a 32-bit binary on a 64-bit architecture)

**Comment by anthony-cros** [comment]
@martinbrook , regarding Jolla, we could try and cross-compile libc-2.17 I suppose, do you think that would work?

**Comment by bobsummerwill** [comment]
No - it's totally not worth it.

@martinbrook By "Jolla", what exactly are you referring to?   A Sailfish OS 1.x release?    I have Sailfish 2.0.0 on my Jolla Phone, for example.    I don't think we necessarily need to worry about older versions so much.    That wouldn't be an issue when we have native builds, and for "generic armhf" cross-builds, I think it's reasonable to have a higher baseline.    We're still to jump everything up to armv7, for example.

**Comment by anthony-cros** [comment]
so the offending instructions that I'm currently blocked on is the following (prettified):
```
/home/xcompiler/x-tools/arm-unknown-linux-gnueabi/bin/arm-unknown-linux-gnueabi-g++ \
 -fPIC \
 -shared \
 -Wl,-soname,libjsonrpccpp-common.so.0 \
 -o \
 ../../lib/libjsonrpccpp-common.so.0.6.0 \
 CMakeFiles/jsonrpccommon.dir/common/errors.cpp.o \
 CMakeFiles/jsonrpccommon.dir/common/specificationwriter.cpp.o \
 CMakeFiles/jsonrpccommon.dir/common/procedure.cpp.o \
 CMakeFiles/jsonrpccommon.dir/common/exception.cpp.o \
 CMakeFiles/jsonrpccommon.dir/common/specificationparser.cpp.o \
 /home/xcompiler/eth/151127041039/installs/jsoncpp/lib/libjsoncpp.a
```
the error message suggests to use the `-fPIC` flag, which is already in use, so not very helpful.

Full message:
```
/home/xcompiler/x-tools/arm-unknown-linux-gnueabi/lib/gcc/arm-unknown-linux-gnueabi/4.9.1/../../../../arm-unknown-linux-gnueabi/bin/ld.bfd: /home/xcompiler/eth/151127041039/installs/jsoncpp/lib/libjsoncpp.a(json_reader.cpp.o): relocation R_ARM_MOVW_ABS_NC against `__pthread_key_create' can not be used when making a shared object; recompile with -fPIC
/home/xcompiler/eth/151127041039/installs/jsoncpp/lib/libjsoncpp.a: could not read symbols: Bad value
collect2: error: ld returned 1 exit status
```
I've tried to switch from `-shared` to `-static` but get an even more unexpected message:

```
/home/xcompiler/x-tools/arm-unknown-linux-gnueabi/arm-unknown-linux-gnueabi/sysroot/usr/lib/crt1.o: In function `_start':
init.c:(.text+0x34): undefined reference to `main'
collect2: error: ld returned 1 exit status
```

I'm unfortunately not very savvy on this kind of matter, anyone has any idea where to go from here? maybe crosstool-ng still isn't configured properly for HF, since that _crt1.o_ file comes from building the toolchain.

**Comment by anthony-cros** [comment]
`readelf -A /home/xcompiler/eth/151127041039/installs/jsoncpp/lib/libjsoncpp.a` --> [gist](https://gist.github.com/anthony-cros/152b69ce02a28c7be6ab)

**Comment by anthony-cros** [comment]
@bobsummerwill , makes sense

> We're still to jump everything up to armv7, for example.
that's already the case in my hacked-up configuration actually, provided it's correct

```
$ readelf -A /home/xcompiler/eth/151116201928/installs/curl/bin/curl | grep 7
  Tag_CPU_name: "7-A"
  Tag_CPU_arch: v7
```

**Comment by bobsummerwill** [comment]
RE:  ARMv7 cool!
But yeah, any glibc mismatches?   Balls to it.    I think that is a jungle which we don't need to enter!

**Comment by bobsummerwill** [comment]
I don't have anything very clever to suggest WRT that -fPIC issue, but I do know what crt1.o is.   That stands for C Runtime Library, and is some glue which calls the user-provided main() function.     There is some startup code in there which initializes the C runtime library, global initializers and so on and then calls main().    That usually means that an APPLICATION is being built, rather than a library or DLL.

So -static is probably wrong.

**Comment by bobsummerwill** [comment]
Oh hang on ... is libjsoncpp.a a prebuilt binary which we're getting from somewhere else?    Or do we built that?    Either way, we're going to need THAT binary to be built with -fPIC too.    I worked through a horror sequence like this on UFC Android.    You have to chase through *all* of your dependencies and get -fPIC binaries for everything.

    /home/xcompiler/eth/151127041039/installs/jsoncpp/lib/libjsoncpp.a

See http://stackoverflow.com/questions/13812185/how-to-recompile-with-fpic

It might be worth downgrading from ARMv7 for the time being just to avoid this issue, which could turn into quite a chase.    Could you try that?    Just back that armv7 out, and see if it goes away.    I think it was added in a later ABI.

**Comment by anthony-cros** [comment]
sounds good, which version do you have in mind? v6?

**Comment by phonikg** [comment]
Just looking at the error from @anthony-cros below...  Does it occur while compiling libjsonrpccpp or libjsoncpp?  If it is libjsonrpccpp, then I have had a few run ins with it not compiling well on armhf.  I have flipped between installing via package manager and compiling on device.

Would it help if I compiled on device in the same manner we are attempting to cross compile to prove the health of this dependency?

/home/xcompiler/x-tools/arm-unknown-linux-gnueabi/bin/arm-unknown-linux-gnueabi-g++ \
 -fPIC \
 -shared \
 -Wl,-soname,libjsonrpccpp-common.so.0 \
 -o \
 ../../lib/libjsonrpccpp-common.so.0.6.0 \

**Comment by anthony-cros** [comment]
>  Does it occur while compiling libjsonrpccpp or libjsoncpp

the former (and it relies on the latter)

**Comment by anthony-cros** [comment]
> Would it help if I compiled on device in the same manner we are attempting to cross compile to prove the health of this dependency?

possibly, would it be difficult?

**Comment by phonikg** [comment]
Nope, I run it in one of my scripts so I can just cut and paste.  I am pulling from this repo:

https://github.com/cinemast/libjson-rpc-cpp

**Comment by anthony-cros** [comment]
yes, that's the one we use too. beware that we have to do some [hacks](https://github.com/doublethinkco/webthree-umbrella/blob/develop/cross-build/ethereum/libjson-rpc-cpp.sh) on our end for it to complete

**Comment by anthony-cros** [comment]
as well as [this one](https://github.com/doublethinkco/webthree-umbrella/blob/develop/cross-build/ethereum/jsoncpp.sh#L48), which is a bit more odd

**Comment by phonikg** [comment]
Hmm... no luck anyway:

/home/debian/libjson-rpc-cpp/build/catch/src/catch/include/internal/catch_tostring.h:56:23: error: ‘nullptr_t’ is not a member of ‘std’
 std::string toString( std::nullptr_t );
                       ^
src/test/CMakeFiles/unit_testsuite.dir/build.make:86: recipe for target 'src/test/CMakeFiles/unit_testsuite.dir/test_server.cpp.o' failed
make[2]: *** [src/test/CMakeFiles/unit_testsuite.dir/test_server.cpp.o] Error 1
CMakeFiles/Makefile2:697: recipe for target 'src/test/CMakeFiles/unit_testsuite.dir/all' failed
make[1]: *** [src/test/CMakeFiles/unit_testsuite.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
debian@debian-armhf:~/libjson-rpc-cpp/build$

**Comment by phonikg** [comment]
Did you see this?

https://github.com/cinemast/libjson-rpc-cpp/issues/135

**Comment by anthony-cros** [comment]
No I hadn't, but that wasn't the error message I was getting though

**Comment by phonikg** [comment]
Yeah I see that, nor is the error I'm getting similar as it seems to point to `catch` for the most part.

**Comment by phonikg** [comment]
I am able to install using package manager with standard linux install instruction, could you switch to using that instead of compiling?

sudo apt-get install libjsonrpccpp-dev libjsonrpccpp-tools -> installs without issue on my armhf device

**Comment by anthony-cros** [comment]
> sudo apt-get install libjsonrpccpp-dev libjsonrpccpp-tools -> installs without issue on my armhf device

that would help get through that one error for the time being, @bobsummerwill , are you ok with that?

**Comment by bobsummerwill** [comment]
The nullptr_t thing is C++11 syntax, I think, and will work for us because of the hack we have.     I don't think we can 'install' because we're running that code on target (not host), so need to cross compile it.    We were able to switch CMake because that was on host for the build process itself

**Comment by bobsummerwill** [comment]
Wrt armv7, I would flip right back to armv5 for now.   That isn't a priority.   We need to get things building and running.  Then look at optimal.

**Comment by bobsummerwill** [comment]
@anthony-cros when you say '-static', where is that setting?

**Comment by anthony-cros** [comment]
the `-static` is in the `g++` command from above, which I had swapped with `-shared` (I also try putting nothing for that flag but got the same error as with `-static`)

**Comment by anthony-cros** [comment]
and sounds good for armv5, trying right now

**Comment by anthony-cros** [comment]
v5 ready:

```
$ readelf -A /home/xcompiler/eth/151127041039/installs/curl/bin/curl
Attribute Section: aeabi
File Attributes
  Tag_CPU_name: "5"
  Tag_CPU_arch: v5T
```

starting the whole process (main.sh) and crossing fingers for libjson-rpc-cpp...

**Comment by anthony-cros** [comment]
it went all the way through (minus an issue I got with cmake but which may just be a problem with my current container rather):

```
$ readelf -A /home/xcompiler/eth/151127224457/installs/webthree/usr/local/bin/eth
Attribute Section: aeabi
File Attributes
  Tag_CPU_name: "5"
  Tag_CPU_arch: v5T
  Tag_ARM_ISA_use: Yes
  Tag_THUMB_ISA_use: Thumb-1
  Tag_FP_arch: VFPv2
  Tag_ABI_PCS_wchar_t: 4
  Tag_ABI_FP_rounding: Needed
  Tag_ABI_FP_denormal: Needed
  Tag_ABI_FP_exceptions: Needed
  Tag_ABI_FP_number_model: IEEE 754
  Tag_ABI_align_needed: 8-byte
  Tag_ABI_align_preserved: 8-byte, except leaf SP
  Tag_ABI_enum_size: int
  Tag_ABI_HardFP_use: SP and DP
  Tag_ABI_VFP_args: VFP registers
  Tag_DIV_use: Not allowed
```
sadly it results in the daunted:
```
sh-4.1$ export LD_LIBRARY_PATH=/opt/usr/media/eth/lib                    
sh-4.1$ ls ./eth
./eth
sh-4.1$ ./eth
sh: ./eth: No such file or directory
```

**Comment by phonikg** [comment]
If you can give me a download link I'll try it on a device...

**Comment by anthony-cros** [comment]
Thanks John, here's a [link](https://drive.google.com/file/d/0B4ByIjYZ3FeJQWhfdEE0NFdwQUU/view?usp=sharing) to it!

**Comment by phonikg** [comment]
./eth: error while loading shared libraries: libboost_thread.so.1.59.0: cannot open shared object file: No such file or directory

**Comment by phonikg** [comment]
Is this a PATH issue?

**Comment by phonikg** [comment]
It Works!!

debian@debian-armhf:/usr/local/bin/eth/bin$ ./eth
(++)Ethereum
Beware. You're entering the Frontier!
...  05:44:27|eth  Reading /home/debian/.web3/keys/eda65840-f872-00cd-6899-38b56ba7ccef.json
⧎ ℹ  05:44:28|eth  Id: ##9733ecbe…

**Comment by anthony-cros** [comment]
Awesome :) yes, you had to set LD_LIBRARY_PATH first! is it sync-ing??

**Comment by phonikg** [comment]
ℹ  06:42:12|eth     Imported and best 25154673110163313  (# 33125 ). Has 0 siblings. Route: [#94b5b516…, #9b3c3da4…]
☍ ℹ  06:42:12|eth     Imported and best 25155908431654300  (# 33126 ). Has 0 siblings. Route: [#9b3c3da4…, #cccb0aaa…]
☍ ℹ  06:42:12|eth     Imported and best 25157144356329608  (# 33127 ). Has 0 siblings. Route: [#cccb0aaa…, #b411d938…]

**Comment by phonikg** [comment]
You can see it here:  https://stats.ethdev.com

Eth(Embedded) - WandBoard Quad ++eth-v1.1.0-0/Debug-Linux/g++/int   18 ms  1	0	#33,133

**Comment by martinbrook** [comment]
did you chmod +x curl ?

**Comment by martinbrook** [comment]
>
> No - it's totally not worth it.
>
@martinbrook <https://github.com/martinbrook> By "Jolla", what exactly are
> you referring to? A Sailfish OS 1.x release? I have Sailfish 2.0.0 on my
> Jolla Phone, for example. I don't think we necessarily need to worry about
> older versions so much. That wouldn't be an issue when we have native
> builds, and for "generic armhf" cross-builds, I think it's reasonable to
> have a higher baseline. We're still to jump everything up to armv7, for
> example.
>

agreed as 2.17 is on old SailfoshOS 1.x release

**Comment by martinbrook** [comment]
>
> so the offending instructions that I'm currently blocked on is the
> following (prettified):
>
> /home/xcompiler/x-tools/arm-unknown-linux-gnueabi/bin/arm-unknown-linux-gnueabi-g++ \
>  -fPIC \
>  -shared \
>  -Wl,-soname,libjsonrpccpp-common.so.0 \
>  -o \
>  ../../lib/libjsonrpccpp-common.so.0.6.0 \
>  CMakeFiles/jsonrpccommon.dir/common/errors.cpp.o \
>  CMakeFiles/jsonrpccommon.dir/common/specificationwriter.cpp.o \
>  CMakeFiles/jsonrpccommon.dir/common/procedure.cpp.o \
>  CMakeFiles/jsonrpccommon.dir/common/exception.cpp.o \
>  CMakeFiles/jsonrpccommon.dir/common/specificationparser.cpp.o \
>  /home/xcompiler/eth/151127041039/installs/jsoncpp/lib/libjsoncpp.a
>
> the error message suggests to use the -fPIC flag, which is already in
> use, so not very helpful.
>
> Full message:
>
> /home/xcompiler/x-tools/arm-unknown-linux-gnueabi/lib/gcc/arm-unknown-linux-gnueabi/4.9.1/../../../../arm-unknown-linux-gnueabi/bin/ld.bfd: /home/xcompiler/eth/151127041039/installs/jsoncpp/lib/libjsoncpp.a(json_reader.cpp.o): relocation R_ARM_MOVW_ABS_NC against `__pthread_key_create' can not be used when making a shared object; recompile with -fPIC
> /home/xcompiler/eth/151127041039/installs/jsoncpp/lib/libjsoncpp.a: could not read symbols: Bad value
> collect2: error: ld returned 1 exit status
>
>
I thin its saying it can't link this library with pthreads as pthreads is
not compiled -fPIC

**Comment by martinbrook** [comment]
>
> Thanks John, here's a link
> <https://drive.google.com/file/d/0B4ByIjYZ3FeJQWhfdEE0NFdwQUU/view?usp=sharing>
> to it!
>
>
> On SailfishOs 2.0

[nemo@Jolla bin]$ ./eth
./eth: /usr/lib/libstdc++.so.6: version `CXXABI_1.3.8' not found (required
by /home/nemo/eth/lib/libleveldb.so.1)
./eth: /usr/lib/libstdc++.so.6: version `GLIBCXX_3.4.20' not found
(required by /home/nemo/eth/lib/libp2p.so)
./eth: /usr/lib/libstdc++.so.6: version `GLIBCXX_3.4.20' not found
(required by /home/nemo/eth/lib/libevmasm.so)
./eth: /usr/lib/libstdc++.so.6: version `GLIBCXX_3.4.20' not found
(required by /home/nemo/eth/lib/liblll.so)
./eth: /usr/lib/libstdc++.so.6: version `GLIBCXX_3.4.20' not found
(required by /home/nemo/eth/lib/libjsonrpccpp-common.so.0)

Sailfish is CXXABI 1.3.3, GLIBC3.4.15 so a similar issue to SailfishOX 1.X
and libc

**Comment by bobsummerwill** [comment]
>> Pushing on with SailfishOS native build

Good job, @martinbrook!   Thanks.

Now we have our first working cross-build (WOOT!), I think we should be able to get that onto the other ARM platforms pretty quickly.   It should just be a matter of getting @anthony-cros's local edits into a form in which they can be committed, and then just tweaking around with these libraries version and ABI settings to get each platform happy in-turn.    If we cannot find a combination of settings which work for all ARM platforms of interest then that is the point where we would have to flip to platform-specific cross-builds (at least for the "fussy" platforms).    Which is fine.

Where we have the option of native builds, that's probably what will ultimately be best, because we'll be using specifically-tailored official toolchains.

This "generic ARM" cross-build toolchain is really intended more as a porting aid than as the ultimate goal, so we should just make it as universal as we are able to, but that can obviously never cover everything.   We'll always have at least armel vs armhf, for example :-)

And then we can look at Intel, but I will break that into its own new Issue.

**Comment by phonikg** [comment]
Success on RasPi2 so far... may still need SWAP:

pi@raspberrypi:~/eth/bin $ ./eth
(++)Ethereum
Beware. You're entering the Frontier!
⧎ ℹ  15:18:00|eth  Id: ##23961e67…

**Comment by bobsummerwill** [comment]
Cool :-)

**Comment by phonikg** [comment]
C.H.I.P. install went well, but then:

chip@chip:~/eth/bin$ ./eth
(++)Ethereum
Beware. You're entering the Frontier!
⧎ ℹ  05:06:36|eth  Id: ##7346fb5b…
pure virtual method called
terminate called without an active exception
Aborted
chip@chip:~/eth/bin$

**Comment by bobsummerwill** [comment]
Thanks for guinea pigging.   Do you have a gdb on the C.H.I.P.?   Are you able to do:

gdb ./eth

?

**Comment by bobsummerwill** [comment]
Lets move the C.H.I.P. discussion to #40.

In fact, I think we're close to being able to close this issue, because we have working armhf builds now, and then will splinter into further specific issues per device, and for Intel support, etc

**Comment by bobsummerwill** [comment]
@anthony-cros Let's close this issue when you have committed your armhf changes and others are able to build their own armhf binaries.

**Comment by bobsummerwill** [comment]
Intel support is #37

**Comment by phonikg** [comment]
FYI, with eth-netstats running in the background it will take > 36 hours to fully sync the blockchain.  Without it running it should take less.

**Comment by bobsummerwill** [comment]
On what, @phonikg?  A native build on chip?

**Comment by phonikg** [comment]
Lol... no.  On the original Wandboard.  If you look on stats.ethdev.com it's still syncing, but almost complete.

Native C.H.I.P. build did not go well.  Have to sort out repo's to install libjsonrpccpp dependencies.

**Comment by bobsummerwill** [comment]
Right.    Yeah - the sync times can be atrocious on some of these low-spec devices.   Roll-on the Light Client.  We so need it :-)

**Comment by bobsummerwill** [comment]
Stumbled across this:

http://unity.ubuntu.com/mir/building_source_for_android.html

Which describes three approaches for building ARM binaries.

The emulated QEMU approach is the route which Damien, who was contracting for me, took for MonoTizen, to build the Mono runtime for Tizen.   See http://monotizen.com.

The third approach described is what we are doing, but with a bonus, which is that it appears that there are prebuilt GCC armhf cross-compiler packages.   We should try those as an alternative to the ones we are currently building ourselves with crosstool-NG.   It shouldn't be too hard to try, I don't think.   We should just apt-get install within the Docker container and then just fix up the 'glue' to point to the right place for those GCC binaries.

**Comment by bobsummerwill** [comment]
Further to the above ...

apt-get install g++-arm-linux-gnueabi
apt-get install g++-arm-linux-gnueabihf

So I'll try that out.

**Comment by anthony-cros** [comment]
> I think its saying it can't link this library with pthreads as pthreads is not compiled -fPIC

thanks @martinbrook , you're probably right actually

**Comment by anthony-cros** [comment]
> getting @anthony-cros's local edits

to be committed soon, it's actually quite small

> tweaking around with these libraries version and ABI settings to get each platform happy in-turn

that's going to be the trickier part, but I'll make a convenience function. I'm hoping to make it open-close-principle friendly so everyone can just extend a baseline script.

> Let's close this issue when you have committed your armhf changes

sounds good

**Comment by anthony-cros** [comment]
> We should try those as an alternative to the ones we are currently building ourselves with crosstool-NG

do you need me to try on the TM1 too?

**Comment by anthony-cros** [comment]
> I'll make a convenience function. I'm hoping to make it open-close-principle friendly so everyone can just extend a baseline script.

==> https://github.com/doublethinkco/webthree-umbrella-cross/commit/210dcbde91b8d879769f85f53f5309f96d6a41c1

**Comment by bobsummerwill** [comment]
>> do you need me to try on the TM1 too?

Try to your cross-built armhf binary on the TM1?   Yes, please!   It might "just work".   #20 is now tracking Tizen support specifically, so that would be the place to put that result.

So to get this working for me, I am going to need to make an extra Dockerfile, right?   Which passes an extra parameter pointing to an armhf config?

**Comment by anthony-cros** [comment]
> Try to your cross-built armhf binary on the TM1?

no i meant the apt-get install g++-arm-linux-gnueabi/apt-get install g++-arm-linux-gnueabihf combination, sadly I already tried the armhf binary i created earlier, but [no dice](https://github.com/doublethinkco/webthree-umbrella-cross/issues/10#issuecomment-160248996)

**Comment by anthony-cros** [comment]
> So to get this working for me, I am going to need to make an extra Dockerfile, right?

so you could reuse the current docker file for that, just pass along the parameter ("default" or "/my/path/to/my/config")

**Comment by bobsummerwill** [comment]
>> so you could reuse the current docker file for that
Right ... if we add some parameterization to the Docker file itself, which I *think* is possible?

**Comment by bobsummerwill** [comment]
@anthony-cros said -  no i meant the apt-get install g++-arm-linux-gnueabi/apt-get install g++-arm-linux-gnueabihf combination, sadly I already tried the armhf binary i created earlier, but no dice

Replied in #42.

See also #20 (Support Tizen).

**Comment by anthony-cros** [comment]
> Right ... if we add some parameterization to the Docker file itself, which I think is possible?

yes, at least with the `-e` flag for EVs, though possibly simply as an argument to docker-build if it's possible

**Comment by anthony-cros** [comment]
See https://github.com/doublethinkco/webthree-umbrella-cross/commit/b99d13a32d9f6ed0f81430a3b205be6acd577fd7 :)

**Comment by bobsummerwill** [comment]
WOOT!    Nice one.

We still need you to upload your armhf xcompiler results, though, right?

Like updating Docker-eth to something like ...

    RUN wget ‐‐output-document=x-tools.tar.gz https://github.com/doublethinkco/webthree-    umbrella-cross/releases/download/0.0.3/x-tools-armel.tar.gz; exit 0
    RUN tar xvfz x-tools-armel.tar.gz
    RUN wget ‐‐output-document=x-tools.tar.gz https://github.com/doublethinkco/webthree-    umbrella-cross/releases/download/0.0.3/x-tools-armhf.tar.gz; exit 0
    RUN tar xvfz x-tools.armhf.tar.gz

**Comment by anthony-cros** [comment]
sounds good, i'm rerunning them and will upload!

**Comment by bobsummerwill** [comment]
Thanks!

I don't understand how YOU were able to run ./build-wandboard.sh yourself.     Wasn't YOUR container missing the xcompiler files for unknown-arm-linux-gnuabihf too?

**Comment by anthony-cros** [comment]
oh no, I only ran the ./build-xcompiler-wandboard.sh one!

**Comment by bobsummerwill** [comment]
Ha ha!
Looks like the wandboard.tgz you uploaded is wrong too, eh?    It's the build scripts, not the xcompiler files, not the eth.

**Comment by anthony-cros** [comment]
hmm i seem to be getting an error on upload, let me try again

**Comment by anthony-cros** [comment]
https://github.com/doublethinkco/webthree-umbrella-cross/releases/tag/wandboard.tgz

**Comment by anthony-cros** [comment]
a bit more work needed here actually

**Comment by anthony-cros** [comment]
see https://github.com/doublethinkco/webthree-umbrella-cross/pull/53 for latest development

I'm getting this however, @bobsummerwill , could it be something you changed recently?

```
(CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/bash /home/crosseth/eth/wd/libmicrohttpd/missing autoheader)
/home/crosseth/eth/wd/libmicrohttpd/missing: line 81: autoheader: command not found
WARNING: 'autoheader' is missing on your system.
         You should only need it if you modified 'acconfig.h' or
         'configure.ac' or m4 files included by 'configure.ac'.
         The 'autoheader' program is part of the GNU Autoconf package:
         <http://www.gnu.org/software/autoconf/>
         It also requires GNU m4 and Perl in order to run:
         <http://www.gnu.org/software/m4/>
         <http://www.perl.org/>
make: *** [MHD_config.h.in] Error 127
```

**Comment by bobsummerwill** [comment]
MERGED!   And will do some runs and see where I can get to :-)

**Comment by bobsummerwill** [comment]
All good.   I'm re-declaring this as done.   Have uploaded new binaries to release.   It's in a state where we can reliably produce armel and armhf binaries now, and start iterating on the ABI details.

**Comment by anthony-cros** [comment]
so it appears the ubuntu touch phone is armhf: `ubuntu build description: ubuntu 15.04 - armhf`, I didn't realize that earlier (I thought it was armel actually)
